[[ Aggregates ]] Add aggregate support to LCB

runrevmark opened the pull request:
This patch contains an implementation of aggregates (e.g. structs) in LCB.

It uses a parameterized typeinfo to allow an aggregate to be defined by a sequence of fields. The aggregate type bridges to/from a list type.

This seems to work as long as the aggregate type is 'public' - non-public types suffer from issues with checking the types at runtime... i.e. the problem with alias vs. named typeinfos which we really need to resolve.

This approach at least gives us the ability to pass and return 'flat' (aggregates which don't use pointers to other aggregates) structs to and from functions. Indeed, by using appropriate 'out' and 'inout' modes, we can also handle functions which take a pointer to struct, or mutate a pointer to a struct (although immense care is needed here!).

runrevmark committed:
[[ Bug 20325 ]] Add notion of foreign type constructor

This patch adds the notion of a 'foreign type constructor'. This
expands the foreign type clause in lcb to allow a binding string of
the form "<name>:<args>". Here, name is resolved as a C function with
prototype bool (*)(MCStringRef, MCTypeInfoRef&). Where the first
argument is passed the args part of the binding string. The function
should create a new typeinfo based on the arguments it is given.

runrevmark committed:
[[ Bug 20325 ]] Add primitive bool type

This patch reintroduces the kMCPrimitiveTypeBool to enable aggregates to
distinguish between a value which should bridge to Boolean and a value which
should bridge to Number (Integer).

runrevmark committed:
[[ Bug 20325 ]] Parameterize foreign type methods with descriptor

This patch parameterizes most of the foreign type methods with a
pointer to the descriptor as the first argument.

runrevmark committed:
[[ Bug 20325 ]] Add aggregate type family

This patch adds a parameterized aggregate type to lcb. The aggregate
type allows simple structs to be defined as a sequence of fields,
with the values bridging to and from List.

livecodeali commented:
@livecode-vulcan review ok f9c8262

livecode-vulcan commented:
💙 review by @livecodeali ok f9c8262eeaaef26f62cebad7d2bce32cd0a86d9b

livecodeali commented:
@runrevmark Looks like you didn't update the foreign type methods in foundation-objc-dummy

runrevmark committed:
[[ Bug 20325 ]] Update signatures of dummy objc foreign types

This patch updates the signatures of the dummy objc foreign types
used on platforms which don't support objc to take into account
the new parameterization.

runrevmark commented:
@livecodeali: Done!

livecodeali commented:
@livecode-vulcan review ok 49497cb

livecode-vulcan commented:
💙 review by @livecodeali ok 49497cb6760ab509102ec32f2bdd414cc31f1e32

livecode-vulcan commented:
😎 test success 49497cb6760ab509102ec32f2bdd414cc31f1e32

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/729)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/745)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/719)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/662)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/664)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/723)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/743)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/728)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/745)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/727)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/738)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/759)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/723)
